sklearn/utils/__init__.py

```diff
@@ -56,46 +58,54 @@ def __init__(self, extra=''):
 
         """
         self.extra = extra
+        self.msg = msg
+        self.update_doc = update_doc
 
     def __call__(self, obj):
-        if isinstance(obj, type):
+        if inspect.isclass(obj):
             return self._decorate_class(obj)
         else:
             return self._decorate_fun(obj)
 
     def _decorate_class(self, cls):
-        msg = "Class %s is deprecated" % cls.__name__
-        if self.extra:
-            msg += "; %s" % self.extra
+        if (self.msg is None) or ('Function ' in self.msg):
+            self.msg = ("Class %s is deprecated; " + self.extra)
```

Comment:
This is confusing: what is the motivation of the `'Function ' in self.msg` test?

Also, why not appending extra systematically when it's not empty?